pkg/kubectl/cmd/set/set_rule.go

```diff
@@ -0,0 +1,598 @@
+
+		# Add the subresource rule to role/clusterrole
+		kubectl set rule policy role foo --resource=rs.extensions/scale --verb=get,list,delete
+
+		# Add the non resource rule to clusterrole
+		kubectl set rule clusterrole test --non-resource-url="*" --verb=get,post,put,delete
+
+		# Print the result (in yaml format) of updating role/clusterrole from a local, without hitting the server
+		kubectl set rule -f path/to/file.yaml --resource=pods --verb=get --local -o yaml`)
+
+	// Valid resource verb list for validation.
+	validResourceVerbs = []string{"*", "get", "delete", "list", "create", "update", "patch", "watch", "proxy", "redirect", "deletecollection", "use", "bind", "impersonate"}
+
+	// Valid non-resource verb list for validation.
+	validNonResourceVerbs = []string{"*", "get", "post", "put", "delete"}
+)
+
+type ResourceOptions struct {
+	Group       string
+	Resource    string
+	SubResource string
+}
+
+// RuleOptions is the start of the data required to perform the operation. As new fields are added, add them here instead of
+// referencing the cmd.Flags
+type RuleOptions struct {
+	resource.FilenameOptions
+
+	Mapper      meta.RESTMapper
+	Typer       runtime.ObjectTyper
+	Infos       []*resource.Info
+	Encoder     runtime.Encoder
+	Out         io.Writer
+	Err         io.Writer
+	Selector    string
+	ShortOutput bool
+	All         bool
+	Local       bool
+	Remove      bool
+	Cmd         *cobra.Command
+
+	resources       []ResourceOptions
+	resourceNames   []string
+	verbs           []string
+	nonResourceURLs []string
+
+	PrintObject func(cmd *cobra.Command, mapper meta.RESTMapper, obj runtime.Object, out io.Writer) error
+}
+
+func NewCmdRule(f cmdutil.Factory, out io.Writer, errOut io.Writer) *cobra.Command {
+	options := &RuleOptions{
+		Out: out,
+		Err: errOut,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "rule (-f FILENAME | TYPE NAME) --verb=verb --resource=resource.group/subresource [--resource-name=resourcename] [--dry-run]",
+		Short:   "Set the Rule on a role/clusterrole",
+		Long:    fmt.Sprintf(ruleLong),
+		Example: ruleExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.Complete(f, cmd, args))
+			cmdutil.CheckErr(options.Validate(f))
+			cmdutil.CheckErr(options.Run(f))
+		},
+	}
+
+	cmdutil.AddPrinterFlags(cmd)
+	usage := "the resource to update the rules"
+	cmdutil.AddFilenameOptionFlags(cmd, &options.FilenameOptions, usage)
+	cmd.Flags().BoolVar(&options.All, "all", false, "select all resources in the namespace of the specified resource types")
+	cmd.Flags().StringVarP(&options.Selector, "selector", "l", "", "Selector (label query) to filter on, supports '=', '==', and '!='.")
+	cmd.Flags().BoolVar(&options.Local, "local", false, "If true, set resources will NOT contact api-server but run locally.")
+	cmd.Flags().BoolVar(&options.Remove, "remove", false, "If true, set resources remove rule from role/clusterrole.")
```

Comment:
There is no precedent for --remove in `kubectl set` commands. Since there are plans to have something like `kubectl unset` I would leave this out from here.